Update advanced-replace extension

Gitsack opened the pull request:
## Description

After some feedback of the community, I worked on adding a second replacement method which solves different needs:

Contrary to the "copy and paste" method which is still available, the new replacement type works like a simpler search & replace function. 

There can be one or more entries which consist of
- a Regex (the search)
- a replacement
- optional checkboxes for regex modifiers (multiline, global, case insensitive)

## Screenshots

![advanced-replace 2025-02-22 at 18 41 42](https://github.com/user-attachments/assets/146df738-77d3-48d9-b177-bc93e376b8ab)

![advanced-replace 2025-02-22 at 18 42 01](https://github.com/user-attachments/assets/df9e42da-4964-4916-bde4-e4f3df86d61e)


<!-- If you add a new extension or command, include a screencast (or screenshot for straightforward changes). A good screencast will make the review much faster - especially if your extension requires registration in other services.  -->

## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)
- [x] I checked that files in the `assets` folder are used by the extension itself
- [x] I checked that assets used by the `README` are placed outside of the `metadata` folder

pernielsentikaer commented:
Hi 👋

Thanks for the update @Gitsack 🔥 

Could you look into the CI errors [mentioned here](https://github.com/raycast/extensions/actions/runs/13475021215/job/37653117388?pr=17236)

@greptileai can you give it a spin too

Gitsack commented on extensions/advanced-replace/src/utilities/replacements.ts:
```diff
@@ -1,12 +1,65 @@
+
+    let output = "";
+
+    entry.regexItems.forEach((e) => {
+      if (!e.regex) {
+        showToast({ title: "Action cancelled", message: "No regex was entered.", style: Toast.Style.Failure });
+        return;
+      }
+
+      if (!e.replacement) {
+        showToast({ title: "Action cancelled", message: "No replacement was entered.", style: Toast.Style.Failure });
+        return;
+      }
+
+      output = content.replace(
+        new RegExp(
+          e.regex,
+          "" + (e.matchGlobally ? "g" : "") + (e.matchCaseInsensitive ? "i" : "") + (e.matchMultiline ? "m" : ""),
+        ),
+        e.replacement,
+      );
+    });
```
Nice catch, thanks botty! :)

Gitsack commented on extensions/advanced-replace/src/components/FormDirectReplace.tsx:
```diff
@@ -0,0 +1,133 @@
+  const [regexItems, setRegexItems] = useState<RegexItemDirectReplace[]>(initialValues?.regexItems || [emptyRegexItem]);
+
+  function addRegexItem() {
+    setRegexItems((prev) => [...prev, emptyRegexItem]);
```
I moved the constant emptyRegexItem to a separate create function 👍

Gitsack commented on extensions/advanced-replace/src/components/FormDirectReplace.tsx:
```diff
@@ -0,0 +1,133 @@
+        style: Toast.Style.Success,
+        title: "Success!",
+        message: `New Regex Option created: ${values.title} (${values.description})`,
+      });
```
I updated the message to be dynamic.

Gitsack commented on extensions/advanced-replace/src/components/FormDirectReplace.tsx:
```diff
@@ -0,0 +1,133 @@
+          regexItems,
+        });
+      } else {
+        const itemIndex = replacementEntries?.findIndex((e) => e.id === initialValues.id);
```
👍 Integrated the extra error handling.

Gitsack commented on extensions/advanced-replace/src/components/FormCutPaste.tsx:
```diff
@@ -0,0 +1,116 @@
+        if (invalidRegexItem) {
+          return "Each Regex item requires both a key and a regex pattern.";
+        }
+      },
```
Changed to explicitly return undefined for a positive validation.

Gitsack commented on extensions/advanced-replace/src/components/FormCutPaste.tsx:
```diff
@@ -0,0 +1,116 @@
+        style: Toast.Style.Success,
+        title: "Success!",
+        message: `New Regex Option created: ${values.title} (${values.description})`,
+      });
```
Was adapted to be dynamic. 👍

Gitsack commented on extensions/advanced-replace/src/components/FormCutPaste.tsx:
```diff
@@ -0,0 +1,116 @@
+          type: "cutPaste",
+          regexItems,
+        });
+      } else {
```
I added the empty fallback array as suggested here.

Gitsack commented on extensions/advanced-replace/src/advanced-replace.tsx:
```diff
@@ -1,12 +1,25 @@
+};
+
+export default function ManageOptions(props: Readonly<LaunchProps<{ draftValues: EntryCutPaste }>>) {
+  const { data: replacementEntries, revalidate } = usePromise(getSavedItems);
```
`isLoading` is now correctly being destructured and passed further down.

Gitsack commented on extensions/advanced-replace/src/components/EntryForm.tsx:
```diff
@@ -0,0 +1,44 @@
+      >
+        <Form.Dropdown.Item value="directReplace" title="Direct Replace" icon={Icon.Bolt} />
+        <Form.Dropdown.Item value="cutPaste" title="Cut Paste" icon={Icon.Code} />
+      </Form.Dropdown>
```
The dropdown component was extracted as a separate component.

Gitsack commented:
@pernielsentikaer thanks for your feedback, I resolved the CI error (which was also mentioned by the AI comments later on).

All feedback suggestions should be resolved now, at least all but one which still is unresolved: As mentioned above, I'm not aware of an option to include a button or a button-like component which can add the suggested behavior to remove a specific replacement option.

Gitsack commented:
@pernielsentikaer ok this might have been some sort of merge issue or git thing, I added an empty `RegexItemForm.tsx` and deleted it afterwards, so CI check step is 🟢.
